File: Leetcode_with_Cpp/77combine.cpp
class Solution {
    vector<vector<int>> res;

public:
    vector<vector<int>> combine(int n, int k) {
        res.clear();
        if(n <= 0 || k <= 0 || n < k)
            return res;
        vector<int> c;
        generateCombine(n, k, 1, c);
        return res;
    }

    void generateCombine(int n, int k, int start, vector<int>& c) {
        if(c.size() == k) {
            res.push_back(c);
            return;
        }
        for(int i = start; i <= n - (k - c.size()) + 1; i++) {  // 剪枝--减少循环次数
            c.push_back(i);
            generateCombine(n, k, i+1, c);
            c.pop_back();
        }
        return;
    }
};

File: Leetcode_with_py/77combine.py
class Solution(object):
    def combine(self, n, k):
        def generateCombine(start, temp):
            if len(temp) == k:
                res.append(temp)
                return
            for i in range(start, n-(k-len(temp))+2):
                generateCombine(i+1, temp+[i])
            return
        if n<=0 or k<=0 or n<k:
            return []
        res = []
        generateCombine(1, [])
        return res

class Solution(object):
    def combine(self, n, k):
        if n<=0 or k<=0 or n<k:
            return []
        # ÅÅÁĞ×éºÏ£º C(m,n) = C(m-1,n) + C(m-1,n-1)
        if k==1:
            return [[i] for i in range(1, n+1)]
        if k==n:
            return [[i for i in range(1, n+1)]]
        res = self.combine(n-1, k)
        for item in self.combine(n-1, k-1):
            item.append(n)
            res.append(item)
        return res